lib/ipam/ipam.go

```diff
@@ -1383,6 +1349,69 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// ensureConsistentAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block affinity was released.
+func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx := log.WithFields(log.Fields{"cidr": b.CIDR, "host": host})
+
+	// If no hostname is found on the block affinity,
+	// there is no need to do an ip pool node selection check.
+	if host == "" {
+		logCtx.Debug("Block already has no affinity")
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node labels for host affinity")
+	node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			logCtx.WithError(err).WithField("node", host).Error("Failed to get node for host")
+			return false, err
+		}
+		logCtx.Info("Node doesn't exist, releasing IPs anyway")
+		return false, nil
+	}
+
+	// Make sure the returned value is a valid node.
+	v3n, ok := node.Value.(*v3.Node)
+	if !ok {
+		return false, fmt.Errorf("Datastore returned malformed node object")
+	}
+
+	// Fetch the pool for the given CIDR and check if it selects the node.
+	if pool := c.blockReaderWriter.getPoolForIP(cnet.IP{b.CIDR.IPNet.IP}); pool == nil {
+		logCtx.Debug("no pools own this block")
+		return false, nil
+	} else {
+		logCtx = logCtx.WithField("pool", pool)
+		if sel, err := pool.SelectsNode(*v3n); err != nil {
+			logCtx.WithError(err).Error("Failed to determine node selection")
+			return false, err
+		} else if sel {
+			logCtx.Debug("Pools selects node, no change")
+			return false, nil
+		} else {
+			logCtx.Debug("Pools no longer selects node, releasing block affinity")
+		}
+	}
+
+	// Pool does not match this node's label, release this block's affinity.
+	if err := c.ReleaseAffinity(ctx, b.CIDR, host); err != nil {
```

Comment:
I think we might be doing too much by calling `c.ReleaseAffinity`.

We could optimize this by copying this logic: https://github.com/projectcalico/libcalico-go/blob/061482258438a87d00299fd86c5780b445997fb7/lib/ipam/ipam.go#L898-L911

`c.ReleaseAffinity` some some stuff we don't need to - e.g., it re-queries the IP pools and checks that the CIDR is within a pool. It also checks the hostname, and implements retries, all of which we're already doing here.

Reply:
Good call, but I'm not a big fan of copy/pasta when we can avoid it. It looks like this code is also duplicated here:

https://github.com/projectcalico/libcalico-go/blob/061482258438a87d00299fd86c5780b445997fb7/lib/ipam/ipam.go#L968

I'll try to break it out as its own function first and if that doesn't work, then I'll start boiling some water 😁